### PR TITLE
Filtering criteria in apt-sm-plugin output

### PR DESCRIPTION
**Pull request by Ruadhri17**
## Proposed changes
This PR adds two features:
- option to filter packages by name and maintainer in apt plugin. User have to set filters manually using `tedge.toml`.
- limit of software list output to given number of lines using `tedge config`. Default value is set to 1000 and 0 means no limit to the output (STC).

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#1926 
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by didier-wenzek** [comment]
Filtering the output of software-package lists must be addressed at two levels.

- At the global level, by the `tedge-agent`, the filtering can be done only on the generic attributes as the number of entries or the package names.
- At the plugin level, in this precise case the apt-plugin, the filtering can be more specific using for instance the package maintainers.

This means for this PR:

- The [change adding a filter parameter to the `Plugin::list()` method](https://github.com/thin-edge/thin-edge.io/pull/1961/files#r1192040608) must be reverted, as this would impact all the plugins while used only by one.
- What has been done to filter out the number of packages per plugin, can be kept as proposed here (I have some disagreements but minor).
- Its a good idea to add  `- --name` and `--maintainer` parameters to the `apt-plugin` command. These parameters will be ignored by the `tedge-agent` but will be useful for testing purposes.
- As the  `- --name` and `--maintainer` parameters are ignored by the `tedge-agent`, one needs to add some configuration for these parameters in `tedge_config.toml`. However, it's not mandatory to have these parameter managed by the `tegde config` command. One can add a specific stanza for the apt-plugin in `tedge_config.toml`. Doing so the apt-plugin will need no dependency on the tedge_config crate and can simply use serde to read the relevant parts of the config.

 @reubenmiller is this okay for you that the apt-plugin read its config from `tedge_config.toml`, but using *free* settings not managed by the `tedge config` command? Such support can be added later and starting with something simpler would make this PR more independent of the tegde-config refactoring.

**Comment by reubenmiller** [comment]
> 
> @reubenmiller is this okay for you that the apt-plugin read its config from `tedge_config.toml`, but using _free_ settings not managed by the `tedge config` command? Such support can be added later and starting with something simpler would make this PR more independent of the tegde-config refactoring.

Yes I'm ok with this proposal for now. We might be a bit silent about the feature/docs until the support of the `tedge config get/set` is supported out of the box using the config refactoring PR.

File: plugins/tedge_apt_plugin/Cargo.toml [hunk 1]
```diff
@@ -27,7 +27,9 @@ clap = { version = "3", features = ["derive"] }
 csv = "1.1"
 regex = "1"
 serde = { version = "1", features = ["derive"] }
+tedge_config = { path = "../../crates/common/tedge_config" }
 thiserror = "1.0"
+toml = "0.5.11"
```
Comment (didier-wenzek) [inline plugins/tedge_apt_plugin/Cargo.toml:32]: We decided to not specify the PATCH number of the dependencies.

```suggestion
toml = "0.5"
```

File: plugins/tedge_apt_plugin/src/main.rs [hunk 2]
```diff
@@ -269,13 +283,27 @@ fn get_name_and_version(line: &str) -> (&str, &str) {
     (name, version)
 }
 
+fn get_config() -> Option<TedgeConfig> {
+    let config_dir = TEdgeConfigLocation::default();
+
+    match fs::read_to_string(config_dir.tedge_config_file_path()) {
+        Ok(content) => {
+            let config: Option<TedgeConfig> = toml::from_str(&content).unwrap_or_else(|e| {
+                eprintln!("Failed to parse TOML: {}", e);
```
Comment (didier-wenzek) [inline plugins/tedge_apt_plugin/src/main.rs:292]: It would be good to tell which file cannot be parsed.

File: plugins/tedge_apt_plugin/src/main.rs [hunk 3]
```diff
@@ -284,6 +312,19 @@ fn main() {
...
+        ref mut name,
+        ref mut maintainer,
+    } = apt.operation
+    {
+        if name.is_none() && maintainer.is_none() {
+            if let Some(config) = get_config() {
+                *name = config.apt.name;
+                *maintainer = config.apt.maintainer;
+            }
+        }
```
Comment (didier-wenzek) [inline plugins/tedge_apt_plugin/src/main.rs:325]: Here the configuration file is used if nothing is provided on the command line. I would do something slightly different, taking the values provided by the configuration file unless explicitly provided on the command line.

File: plugins/tedge_apt_plugin/src/main.rs [hunk 4]
```diff
@@ -70,6 +73,17 @@ struct SoftwareModuleUpdate {
...
+#[derive(Debug, Deserialize)]
+struct TedgeConfig {
+    pub apt: AptConfig,
+}
+
+#[derive(Debug, Deserialize)]
+struct AptConfig {
+    pub name: Option<String>,
+    pub maintainer: Option<String>,
+}
```
Comment (didier-wenzek) [inline plugins/tedge_apt_plugin/src/main.rs:85]: That's perfect. We only need these data from the configuration file and we don't need to use tedge-config to get those.

**Review by didier-wenzek — COMMENTED** [review]
Nice, things are going in the expected direction.

File: plugins/tedge_apt_plugin/src/main.rs [hunk 5]
```diff
@@ -64,27 +73,56 @@ struct SoftwareModuleUpdate {
...
+
+                _ => Regex::new(
+                    format!(
+                        r"(^{}\t.*|^\S+\t\S+\t{}\s+.*)install ok installed",
+                        name.unwrap_or_default(),
+                        maintainer.unwrap_or_default()
+                    )
+                    .as_str(),
+                )
+                .unwrap(),
```
Comment (PradeepKiruvale) [inline plugins/tedge_apt_plugin/src/main.rs:115]: remove unwrap

File: plugins/tedge_apt_plugin/src/main.rs (hunk 2 above)
Comment (didier-wenzek) [reply plugins/tedge_apt_plugin/src/main.rs]: It would be better to give the path actually used:

```
                eprintln!("Failed to parse {}: {}", config_dir.tedge_config_file_path(), e);
```

File: plugins/tedge_apt_plugin/src/main.rs [hunk 6]
```diff
@@ -64,27 +73,56 @@ struct SoftwareModuleUpdate {
...
-            for line in stdout.trim_end().split('\n') {
-                if line == "Listing..." {
-                    // Ignore the first line of the stdout
-                    continue;
+                .map_err(|err| InternalError::exec_error("dpkg-query", err))?;
+
+            let stdout = String::from_utf8(dpkg_query.stdout).unwrap_or_default();
+
+            let filter = match (&name, &maintainer) {
+                (None, None) => Regex::new(r"install ok installed").unwrap(),
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply plugins/tedge_apt_plugin/src/main.rs]: This unwrap is safe.

File: plugins/tedge_apt_plugin/src/main.rs (hunk 5 above)
Comment (didier-wenzek) [reply plugins/tedge_apt_plugin/src/main.rs]: This unwrap must be replaced by a proper error handler as the regex is built from user inputs: the filters on package names and maintainers.

File: docs/src/howto-guides/017_apama_software_management_plugin.md [hunk 7]
```diff
@@ -88,3 +88,15 @@ TedgeDemoMonitor                                             1
 ```
 
 You can find more information on this test Apama project [here](https://github.com/thin-edge/thin-edge.io_examples/tree/main/StreamingAnalytics#testing-a-project).
+
+## Filter packages by name and maintainer
+
+You can filter the list output using two filtering criteria: name and maintainer. Create the `apt` table in `tedge.toml` and fill it with the `name` and `maintainer` keys. The value of each filter key should be a **valid regex pattern**. Your filters should look like this:
```
Comment (PradeepKiruvale) [inline docs/src/howto-guides/017_apama_software_management_plugin.md:94]:
```suggestion
You can filter the package list output using two filtering criteria: **name and maintainer**. Create the `apt` table in `tedge.toml` and fill it with the `name` and `maintainer` keys. The value of each filter key should be a **valid regex pattern**. Your filters should look like this:
```

File: docs/src/howto-guides/017_apama_software_management_plugin.md [hunk 8]
```diff
@@ -88,3 +88,15 @@ TedgeDemoMonitor                                             1
...
+
+```toml
+[apt]
+name = "exemplary_name.*"
+maintainer  = "exemplary_maintainer"
+```
```
Comment (PradeepKiruvale) [inline docs/src/howto-guides/017_apama_software_management_plugin.md:100]: Can't these be set/unset using the `tedge cli`?

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
The happy path is working as expected.

Two points musty be changed:

- Having no `apt` entry in `tedge.toml` is not an error. The command should proceed with a warning.
- If an incorrect regex is provided as a `name` or `maintainer` filter, the command must  not panic but stop with the appropriate error message.

The document must be move to a more appropriate place.

File: docs/src/howto-guides/017_apama_software_management_plugin.md (hunk 8 above)
Comment (didier-wenzek) [reply docs/src/howto-guides/017_apama_software_management_plugin.md:100]: No this will be addressed once merged the new tedge-config PR.

File: plugins/tedge_apt_plugin/src/main.rs [hunk 9]
```diff
@@ -238,33 +276,34 @@ fn run_cmd(cmd: &str, args: &str) -> Result<ExitStatus, InternalError> {
...
+
+    let name = vec.first().unwrap_or(&"unknown name");
+    let version = vec.get(1).unwrap_or(&"unknown version");
     (name, version)
 }
 
+fn get_config() -> Option<TedgeConfig> {
+    let config_dir = TEdgeConfigLocation::default();
+
+    match fs::read_to_string(config_dir.tedge_config_file_path()) {
```
Comment (albinsuresh) [inline plugins/tedge_apt_plugin/src/main.rs:289]: Wondering why we chose to parse the tedge.toml file directly instead of using the `TEdgeConfig` APIs. To avoid an extra dependency for the plugin?

File: plugins/tedge_apt_plugin/src/main.rs [hunk 10]
```diff
@@ -300,15 +355,10 @@ mod tests {
     use test_case::test_case;
 
     #[test_case(
```
Comment (albinsuresh) [inline plugins/tedge_apt_plugin/src/main.rs:357]: Could you add a test for empty `name` and `maintainer` values as well?

File: crates/common/tedge_config/src/tedge_config_cli/settings.rs [hunk 11]
```diff
@@ -593,3 +593,17 @@ impl ConfigSetting for ServiceTypeSetting {
...
+impl ConfigSetting for SoftwarePluginMaxPackagesSetting {
+    const KEY: &'static str = "software.plugin.max_packages";
+
+    const DESCRIPTION: &'static str = concat!(
+        "The maximum number of software packages reported by each sm-plugin",
+        "Note: If the value is set to 0, there is no limit to reported packages",
+        "Example: 100"
+    );
+
+    type Value = Quantity;
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/common/tedge_config/src/tedge_config_cli/settings.rs:608]: Yeah. Introducing the whole `quantity` module seems a bit heavy for just a numeric value.

File: docs/src/howto-guides/017_apama_software_management_plugin.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply docs/src/howto-guides/017_apama_software_management_plugin.md:100]: In that case, the request is from me. To avoid dependencies with the work not-yet merged on the new tedge config.

File: plugins/tedge_apt_plugin/src/main.rs [hunk 12]
```diff
@@ -64,27 +73,56 @@ struct SoftwareModuleUpdate {
...
+            let stdout = String::from_utf8(dpkg_query.stdout).unwrap_or_default();
+
+            let filter = match (&name, &maintainer) {
+                (None, None) => Regex::new(r"install ok installed"),
+
+                _ => Regex::new(
+                    format!(
+                        r"(^{}\t.*|^\S+\t\S+\t{}\s+.*)install ok installed",
+                        name.unwrap_or_default(),
+                        maintainer.unwrap_or_default()
```
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply plugins/tedge_apt_plugin/src/main.rs:111]: In current scenario, if both name and maintainer are `None` value, the second pattern matching will apply. Basically, I unwrap them to default to fit all possible scenarios in one pattern (If those values unwrapped to `\S+`, you would get whole list even though you provided only one parameter)

File: docs/src/howto-guides/017_apama_software_management_plugin.md [hunk 13]
```diff
@@ -88,3 +88,15 @@ TedgeDemoMonitor                                             1
...
+
+You can filter the list output using two filtering criteria: name and maintainer. Create the `apt` table in `tedge.toml` and fill it with the `name` and `maintainer` keys. The value of each filter key should be a **valid regex pattern**. Your filters should look like this:
+
+```toml
+[apt]
+name = "exemplary_name.*"
+maintainer  = "exemplary_maintainer"
+```  
+
+Also, filters can be provided as a command line parameter in the `apt-plugin`. However, they are created for testing purposes only and will overwrite config parameters.
```
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply docs/src/howto-guides/017_apama_software_management_plugin.md:102]: You are right, it should be override instead of overwrite.

File: plugins/tedge_apt_plugin/src/main.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Ruadhri17) [reply plugins/tedge_apt_plugin/src/main.rs:112]: but if you ask about the case when both values are empty strings, then the function will produce empty list.

File: plugins/tedge_apt_plugin/src/main.rs (hunk 9 above)
Comment (albinsuresh) [reply plugins/tedge_apt_plugin/src/main.rs:294]: Got the answer from @didier-wenzek 's previous review comments.

File: plugins/tedge_apt_plugin/src/main.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply plugins/tedge_apt_plugin/src/main.rs:112]: Yeah, I was talking about empty `""` values only. Initially I was thinking that empty values should behave similar to `None` values. But, seeing your last comment, I agree that empty list is probably even more appropriate. This can be used as a way to filter everything out.

File: plugins/tedge_apt_plugin/src/main.rs [hunk 14]
```diff
@@ -300,18 +364,22 @@ mod tests {
...
-    ; "upgradable"
-    )]
     fn get_package_name_and_version(line: &str, expected_name: &str, expected_version: &str) {
         let (name, version) = get_name_and_version(line);
         assert_eq!(name, expected_name);
         assert_eq!(version, expected_version);
     }
+
+    #[test]
+    fn both_filters_are_empty_strings() {
```
Comment (albinsuresh) [inline plugins/tedge_apt_plugin/src/main.rs:378]: Thanks for adding that. Now, we need test coverage for the following cases as well:

1. name: Some(...), maintainer: Some(...)
2. name: None, maintainer: Some(...)
3. name: Some(...), maintainer: None
4. name: None, maintainer: None
5. name: Some(""), maintainer: Some("")
6. name: Some(...), maintainer: Some("")
7. name: Some(""), maintainer: Some(...)

Since the filtering itself can't be tested with unit tests here, you probably will have to rely on some integration tests as you really need to execute this plugin and parse the output. Here is a test with which you can get started: https://github.com/thin-edge/thin-edge.io/tree/main/tests/RobotFramework/tests/plugin_apt

**Comment by didier-wenzek** [comment]
Config settings have now to be added to `crates/common/tedge_config/src/tedge_config_cli/new.rs`.

I acknowledge that the current status is confusing as there are still some pending cleaning tasks to finalize https://github.com/thin-edge/thin-edge.io/pull/1936.

___=> I will take the time tomorrow to help you and see what needs to be cleaned.__

```
error[E0609]: no field `max_packages` on type `TEdgeConfigDtoSoftware`
    --> crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:1072:14
     |
1072 |             .max_packages
     |              ^^^^^^^^^^^^ unknown field
     |
     = note: available fields are: `plugin`

error[E0609]: no field `max_packages` on type `TEdgeConfigDtoSoftware`
    --> crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:1082:28
     |
1082 |         self.data.software.max_packages = Some(value.into());
     |                            ^^^^^^^^^^^^ unknown field
     |
     = note: available fields are: `plugin`

error[E0609]: no field `max_packages` on type `TEdgeConfigDtoSoftware`
    --> crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:1087:28
     |
1087 |         self.data.software.max_packages = Some(self.config_defaults.default_max_packages.into());
     |                            ^^^^^^^^^^^^ unknown field
     |
     = note: available fields are: `plugin`
```

**Comment by Ruadhri17** [comment]
> Config settings have now to be added to `crates/common/tedge_config/src/tedge_config_cli/new.rs`.

I fixed that issue just before you posted the comment.

If we are after the tedge config refactor, should we now add one for name and maintainer filters?

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config_defaults.rs [hunk 15]
```diff
@@ -90,8 +92,12 @@ pub struct TEdgeConfigDefaults {
 
     /// Default service type
     pub default_service_type: String,
+
     /// Default lock files bool
     pub default_lock_files: Flag,
+
+    /// Default max number of packages
+    pub default_max_packages: Quantity,
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/tedge_config_defaults.rs:100]: Something is wrong here, but beyond the scope of this PR: this default value is not used by `tedge config get software.plugin.max_packages`. Indeed, since https://github.com/thin-edge/thin-edge.io/pull/1936 has been merged, the tedge config command is using the defaults defined in [`tedge_config/.../new.rs`](https://github.com/thin-edge/thin-edge.io/blob/main/crates/common/tedge_config/src/tedge_config_cli/new.rs)

File: crates/common/tedge_config/src/tedge_config_cli/new.rs [hunk 16]
```diff
@@ -460,6 +462,10 @@ define_tedge_config! {
             /// The default software plugin to be used for software management on the device
             #[tedge_config(example = "apt")]
             default: String,
+
+            /// The maximum number of software packages reported by sm-plugin
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/new.rs:466]: This message is for the end-user who is unaware of any `sm-plugin`.

```suggestion
            /// The maximum number of software packages reported for each type of software package
```

File: crates/common/tedge_config/src/tedge_config_cli/new.rs [hunk 17]
```diff
@@ -460,6 +462,10 @@ define_tedge_config! {
...
+            #[tedge_config(example = "1000")]
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/new.rs:467]: A default value of `1000` would make things easier.

File: plugins/tedge_apt_plugin/src/main.rs [hunk 18]
```diff
@@ -64,27 +74,60 @@ struct SoftwareModuleUpdate {
...
+                        maintainer.unwrap_or_default()
+                    )
+                    .as_str(),
+                )
+                .map_err(|err| {
+                    eprintln!("{err}");
+                    std::process::exit(1)
+                }),
+            }
+            .unwrap();
```
Comment (didier-wenzek) [inline plugins/tedge_apt_plugin/src/main.rs:121]: This `unwrap()` is now safe (as there is `map_err()` that exits) - but this is not obvious.

I would use a less error-prone pattern. Say `if let Ok(filter)` or even better a simple `match` that handles the error case (instead of a inner `map_err`) along the ok case.

File: plugins/tedge_apt_plugin/src/main.rs [hunk 19]
```diff
@@ -64,27 +74,60 @@ struct SoftwareModuleUpdate {
...
+                    format!(
+                        r"(^{}\t.*|^\S+\t\S+\t{}\s+.*)install ok installed",
+                        name.unwrap_or_default(),
+                        maintainer.unwrap_or_default()
+                    )
+                    .as_str(),
+                )
+                .map_err(|err| {
+                    eprintln!("{err}");
+                    std::process::exit(1)
```
Comment (didier-wenzek) [inline plugins/tedge_apt_plugin/src/main.rs:118]: This error message is nicer now, but it would be perfect with some context.

```
target/debug/tedge-apt-plugin list --name '[' 
regex parse error:
    (^[\t.*|^\S+\t\S+\t\s+.*)install ok installed
      ^
error: unclosed character class
```

```suggestion
                    eprintln!("tedge-apt-plugin fails to list packages with matching name and maintener: {err}");
                    std::process::exit(1)
```

**Review by didier-wenzek — COMMENTED** [review]
[EDITED] as I changed my mind after a small experiment: it's super easy to use the new tedge config right now.

The main feature is working fine, but some improvements are required before merging.
- The source of truth is now `tedge_config/../new.rs` and the old settings must no more be used.
- The Quantity type can simply be removed as not really useful.
- `max_packages` doc and default value have to be fixed in the new `tedge_config`.
- error handling for ill-formed `name` or `maintainer` settings can be improved.

File: crates/core/plugin_sm/src/plugin.rs [hunk 20]
```diff
@@ -436,10 +438,29 @@ impl Plugin for ExternalPluginCommand {
         let command = self.command(LIST, None)?;
         let output = self.execute(command, logger).await?;
         if output.status.success() {
-            Ok(deserialize_module_info(
-                self.name.clone(),
-                output.stdout.as_slice(),
-            )?)
+            let config = get_tedge_config().unwrap();
```
Comment (rina23q) [inline crates/core/plugin_sm/src/plugin.rs:441]: I don't think this `unwrap()` is safe because if `/etc/tedge/tedge.toml` doesn't exist or is malformed, it should panic.

File: crates/core/plugin_sm/src/plugin.rs [hunk 21]
```diff
@@ -436,10 +438,29 @@ impl Plugin for ExternalPluginCommand {
...
+            let limit = config.query(SoftwarePluginMaxPackagesSetting).unwrap().0;
```
Comment (rina23q) [inline crates/core/plugin_sm/src/plugin.rs:442]: This `unwrap()` is safe because `config.query(SoftwarePluginMaxPackagesSetting)` returns always `Ok`. However, it is always recommended to add a comment why it's safe when you use `unwrap()`.

File: tests/RobotFramework/tests/plugin_apt/filter_packages_list_output.robot [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+
+Apply maintainer filter
+    ${packages_list}=    Execute Command
+    ...    sudo /etc/tedge/sm-plugins/apt list --maintainer thin-edge.*
+    ...    exp_exit_code=0
+    Should Match Regexp    ${packages_list}    c8y.*${VERSION}(.|\\n)*tedge\\s+${VERSION}
+
+Apply both filters
+    ${packages_list}=    Execute Command
+    ...    sudo /etc/tedge/sm-plugins/apt list --name sudo --maintainer thin-edge.*
```
Comment (rina23q) [inline tests/RobotFramework/tests/plugin_apt/filter_packages_list_output.robot:26]: Why the name is `sudo`? Using it as an example package name which doesn't exist?

**Review by rina23q — COMMENTED** [review]
Suddenly dropping in and left some comments like ninja 🥷

File: crates/common/tedge_config/src/tedge_config_cli/models/quantity.rs
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/models/quantity.rs:1]: This file can safely be removed. The new tedge config nicely handles settings with a type of `u32`.

File: crates/common/tedge_config/src/tedge_config_cli/models/mod.rs
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/models/mod.rs:1]: This change can be reverted.

File: crates/common/tedge_config/src/tedge_config_cli/new.rs [hunk 23]
```diff
@@ -460,6 +462,10 @@ define_tedge_config! {
...
+            max_packages: Quantity
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/new.rs:468]: Simply use `u32`

File: crates/common/tedge_config/src/tedge_config_cli/settings.rs
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/settings.rs:1]: The changes on this file have to be reverted: the Setting concept will be removed soon.

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:1]: The changes on this file have to be reverted: the Setting concept will be removed soon.

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config_defaults.rs
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/tedge_config_defaults.rs:1]: The changes on this file have to be reverted: the Setting concept will be removed soon.

File: crates/common/tedge_config/tests/test_tedge_config.rs
Comment (didier-wenzek) [inline crates/common/tedge_config/tests/test_tedge_config.rs:1]: The changes on this file have to be reverted: the Setting concept will be removed soon.

File: crates/core/tedge/src/cli/config/config_key.rs
Comment (didier-wenzek) [inline crates/core/tedge/src/cli/config/config_key.rs:1]: The changes on this file have to be reverted: the Setting concept will be removed soon.

File: crates/core/plugin_sm/src/plugin.rs (hunk 21 above)
Comment (didier-wenzek) [inline crates/core/plugin_sm/src/plugin.rs:442]: To use the new tedge_config is super easy:


```suggestion
            let config = get_new_tedge_config().unwrap();
            let limit = config.software.plugin.max_packages;
```

With just one caveat: the function `get_new_tedge_config()` is not existing yet ;-). It has to be added as a temporary dup of `get_tedge_config()`:

```
/// loads the new tedge config from system default
pub fn get_new_tedge_config() -> Result<new::TEdgeConfig, TEdgeConfigError> {
    let tedge_config_location = TEdgeConfigLocation::default();
    TEdgeConfigRepository::new(tedge_config_location).load_new()
}
```

**Review by didier-wenzek — COMMENTED** [review]
Here the list of changes to make this PR using the new tedge config.

File: crates/core/plugin_sm/src/plugin.rs (hunk 21 above)
Comment (didier-wenzek) [reply crates/core/plugin_sm/src/plugin.rs:442]: I agree with @rina23q comment: using `unwrap()` here is unsafe and also buggy as the config might be stored in a non-default location. The correct way to implement that would be to do the same as for the default plugin type: to let the agent read the config and provide the `TEdgeConfig` value to the plugin. But as this PR is open since quite a long time, I'm okay to have this done in a following PR.

**Review by didier-wenzek — APPROVED** [review]
Approved.

Follow-up tasks:
- The `ExternalPluginCommand::list()` method must not assume the config is at the default location.
- Consider to manage the `name` and `maintainer` filters with `tedge config`.

File: tests/RobotFramework/tests/plugin_apt/filter_packages_list_output.robot (hunk 22 above)
Comment (Ruadhri17) [reply tests/RobotFramework/tests/plugin_apt/filter_packages_list_output.robot:26]: Actually, quite the opposite. Filters work here like union so I wanted to show that when we use it with the maintainer filter set to thin-edge, it will print not only all the tedge packages but also sudo package. And I think we can safely assume that the sudo package will be available (In fact, it is even used when executing the test command).
